protocol/msgpack/base_iterator.go

```diff
@@ -258,3 +268,19 @@ func (it *baseIterator) checkNumFieldsForTypeWithActual(
 	}
 	return numExpectedFields, numActualFields, true
 }
+
+// bufReader is a buffered reader.
```

Comment:
Are we doing this so we can control the size of the `bufio.Reader`? Seems `msgpack` does the same thing already, just without an explicit size (from `decode.go`):

```
type bufReader interface {
	Read([]byte) (int, error)
	ReadByte() (byte, error)
	UnreadByte() error
}

func newBufReader(r io.Reader) bufReader {
	if br, ok := r.(bufReader); ok {
		return br
	}
	return bufio.NewReader(r)
}
...

func NewDecoder(r io.Reader) *Decoder {
	return &Decoder{
		DecodeMapFunc: decodeMap,

		r:   newBufReader(r),
		buf: makeBuffer(),
	}
}
```